chd-rs/src/compression/zstd.rs

```diff
@@ -31,79 +29,67 @@ pub struct ZstdCodec {
 /// when decompressed.
 #[cfg(feature = "fast_zstd")]
 pub struct ZstdCodec {
-    buffer: Vec<u8>,
+    zstd_context: zstd_safe::DCtx<'static>
 }
 
 #[cfg(not(feature = "fast_zstd"))]
 impl CodecImplementation for ZstdCodec {
-    fn new(hunk_size: u32) -> crate::Result<Self>
+    fn new(_hunk_size: u32) -> crate::Result<Self>
     where
         Self: Sized,
     {
         Ok(Self {
-            engine: ruzstd::FrameDecoder::new(),
-            buffer: Vec::with_capacity(hunk_size as usize),
+            decoder: ruzstd::FrameDecoder::new(),
         })
     }
 
-    fn decompress(&mut self, input: &[u8], output: &mut [u8]) -> crate::Result<DecompressResult> {
-        self.buffer.clear();
-        let engine = std::mem::take(&mut self.engine);
-        let mut engine = ruzstd::StreamingDecoder::new_with_decoder(input, engine)
-            .map_err(|_| Error::CodecError)?;
-
-        // If each chunk doesn't output to exactly the same then it's an error
-        let bytes_read = engine
-            .read_to_end(&mut self.buffer)
+    fn decompress(&mut self, mut input: &[u8], output: &mut [u8]) -> crate::Result<DecompressResult> {
+        self.decoder.reset(&mut input).unwrap();
```

Comment:
This should be fallible rather than panic, would just `map_err` it to `DecompressionError`.

Reply:
Oh whoops, I must've done that for testing and then overlooked it.